### DISCUSSION TITLE
noob question - how can I securely connect RDS using node-mysql2?

### DISCUSSION THREAD

**Discussion by Nithur-M** [sidorares/node-mysql2#1742] [Q&A]
I know connecting via the provided endpoint is not safe since it will open up connections to the whole internet. What is the best way to connect to the db on RDS to my node app hosted somewhere else? 
I see using SSH is an option but node-mysql2 don't have any ability to connect that way. Please help. Thanks.

**Comment by sidorares** [comment]
the easiest way is to use ssl encrypted connection that mysql protocol [suports](https://dev.mysql.com/doc/dev/mysql-server/latest/page_protocol_connection_phase.html#sect_protocol_connection_phase_initial_handshake_ssl_handshake)

In order to enable it, you pas `ssl` key as a `createConnection` option ( or as a parameter in a connection url ), documentation is [here](https://github.com/mysqljs/mysql#ssl-options)

In case of AWS RDS you can use a string value as an alias to bundled ssl options configuration, example:

```js
const conn = mysql.createConnection({
  // username, password, etc
  ssl: "Amazon RDS"
});
```

another option is to have ssh server on the same box as mysql client and tunnel it, for example using ssh2 library
see example code here: https://stackoverflow.com/a/30669454/705115

**Comment by Nithur-M** [comment]
Hey @sidorares, thanks for the help. 
For the first method, I still have to open the connection for the internet right? to be able to be accessed from anywhere? is that safe?

**Reply by sidorares** [reply]
yes, db ip/port needs to be accessible for your client source ip range. With ssl the communication is encrypted, and there are currently no known vulnerabilities for `mysql_native_password` or `mysql_caching_sha2_password` auth methods afaik but please do your own research, thats just my opinion :) Companies like https://planetscale.com/ offer hosted db solution where connecting to a db over the internet is a recommended practice so it's probably OK

**Comment by Nithur-M** [comment]
hello @sidorares, is it possible to use promise wrapper in this method?
> another option is to have ssh server on the same box as mysql client and tunnel it, for example using ssh2 library
see example code here: https://stackoverflow.com/a/30669454/705115

**Reply by sidorares** [reply]
yes, promise `createConnection` takes exactly same parameters as a non-promise https://github.com/sidorares/node-mysql2/blob/afaa0484749c8e33243fbd9c9d3e0abfd017bf63/promise.js#L251

so you can pass `stream` to it

**Reply by Nithur-M** [reply]
`const connection = mysql.createConnection(updatedDbServer);
              connection.connect((error) => {
                  if (error) {
                      console.log("error---", error);
                      reject(error);
                  }
                  console.log("Connection Successful");
                  resolve(connection);
              });`
When I pass stream to the createConnection (promise wrapper) function, it is throwing an error connection.connect is not a function. Why?

**Reply by sidorares** [reply]
promise `createConnection` does not have `.connect()` on it and resolves only once connection is established

**Reply by Nithur-M** [reply]
@sidorares Thanks for the help.
